library/bignum.c

```diff
@@ -1796,175 +1639,68 @@ int mbedtls_mpi_exp_mod(mbedtls_mpi *X, const mbedtls_mpi *A,
     }
 
     /*
-     * W[1] = A * R^2 * R^-1 mod N = A * R mod N
+     * Ensure that the exponent that we are passing to the core is not NULL.
      */
-    if (mbedtls_mpi_cmp_mpi(A, N) >= 0) {
-        MBEDTLS_MPI_CHK(mbedtls_mpi_mod_mpi(&W[1], A, N));
-        /* This should be a no-op because W[1] is already that large before
-         * mbedtls_mpi_mod_mpi(), but it's necessary to avoid an overflow
-         * in mpi_montmul() below, so let's make sure. */
-        MBEDTLS_MPI_CHK(mbedtls_mpi_grow(&W[1], N->n + 1));
+    if (E->n == 0) {
+        mbedtls_mpi_lset(&E_core, 0);
     } else {
-        MBEDTLS_MPI_CHK(mbedtls_mpi_copy(&W[1], A));
+        memcpy(&E_core, E, sizeof(mbedtls_mpi));
```

Comment:
I was thinking about that too. I kept them because I thought that the goal is to make the reader stop and think and realise that this is an intentional shallow copy and that is why I kept it. But I too prefer assignment over memcpy here so I will change it.